### ISSUE TITLE
cabal caches invalid store-dir paths

### ISSUE THREAD

**Issue by AndreasPK** [haskell/cabal#5504]
```
PS C:\ghc\msys64\home\Andi\lib_bench\vector\c_allCalls> cabal --store-dir=/c/store-old-l new-run --builddir=d-old-l --project-file=old.project algorithms -j1
Build profile: -w ghc-8.7.20180802 -O1
In order, the following will be built (use -v for more details):
... <package list> ...
Configuring library for Only-0.1..
cabal.exe: expected an absolute directory name for --prefix:
/c/store-old-l\ghc-8.7.20180802\Only-0.1-1612b60b5abb04c23354903aed5a65e522bcda69

cabal.exe: Failed to build Only-0.1 (which is required by
vector-benchmarks-0.10.10). The failure occurred during the configure step.



PS C:\ghc\msys64\home\Andi\lib_bench\vector\c_allCalls> cabal --store-dir=c:\\store-old-l new-run --builddir=d-old-l --project-file=old.project algorithms -j1
Build profile: -w ghc-8.7.20180802 -O1
In order, the following will be built (use -v for more details):
... <package list> ...
Configuring library for Only-0.1..
cabal.exe: expected an absolute directory name for --prefix:
/c/store-old-l\ghc-8.7.20180802\Only-0.1-1612b60b5abb04c23354903aed5a65e522bcda69

cabal.exe: Failed to build Only-0.1 (which is required by
vector-benchmarks-0.10.10). The failure occurred during the configure step
```

The error persists until I delete the given build directory.

**Comment by hvr** [comment]
More generally, cabal doesn't detect changes of the `--store-dir`  value

**Comment by Mistuke** [comment]
I wonder why cabal found the `--store-dir` anyway. As those two paths aren't the same. Cabal correctly recognizes that `/c/store-old-1` is a relative path on Windows[1]. Which in the first case is resolved to `c:\c\store-old-1` due to his active drive being `C:` and the second store dir is `c:\store-old-1`.

So they shouldn't have been the same to begin with[2]... So there are two bugs here.

[1] https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file#fully-qualified-vs-relative-paths
[2] https://blogs.msdn.microsoft.com/jeremykuhne/2016/04/21/path-normalization/

**Comment by hvr** [comment]
potential fix available in #5506

**Comment by hvr** [comment]
cherry-picked to 2.4 via 618efe6
